### PR TITLE
Job Writing Guide

### LINKED ISSUE #438: Explain Operations Properly
The core concept of an Operation is really important in OpenFn scripting, but it's not really well explained anywhere. As a result, I see lots of misconceptions around this in job code.

* An operation is a function which takes state and returns state. 
* Although weirdly, job writers don't see this - they just call a function with an argument and all the state stuff is hidden
* Operations should only be called at the top level of a script

This is correct use of operations:
```
get('/some-data')
post('some-data, (state) => state.data)
```

This is incorrect (and common):
```
get('/some-data', (state) => {
   post('some-data, state.data)(state)
})
```

### PR DESCRIPTION
**Pull request by josephjclark**
Added a new docs page for a guide to job writing.

This is really important for onboarding new users and establishing best practice going forward.

Some of this stuff is mentioned across various pages, some of it is implied, some of it isn't mentioned it all. I would like a single source of truth as both a guide and a reference.

I expect this page to become a source of truth and to be updated often and aggressively in response to user feedback.

Expandable sections add context and should ensure the documentation is useful for different levels of ability.

This should close #438

### CODE REVIEW

**Comment by aleksa-krolls** [comment]
@mtuchi can you pls review this today?

File: docs/job-writing-guide.md [hunk 1]
```diff
@@ -0,0 +1,750 @@
...
+
+````
+get('/patients')
+```js
+Compiles to this Javascript module:
+```js
+import { get } from "@openfn/language-http";
+export * from "@openfn/language-http";
+export default [get('/patients')];
+````
```
Comment (aleksa-krolls) [inline docs/job-writing-guide.md:750]: @josephjclark at the end, I think we should direct folks to where to go next/learn more... maybe to the main [job docs page](https://docs.openfn.org/documentation/build/steps/jobs)? 
And how about a call to action, such as asking folks to post questions on Community if they get stuck?

File: docs/job-writing-guide.md [hunk 2]
```diff
@@ -0,0 +1,750 @@
...
+This guide applies equally to writing Jobs on the app (Lightning) or through the
+CLI.
+
+:::info Workflows Multiple jobs can be chained together in a Workflow. A common
+pattern is to use one job to fetch data from datasource A, one job to convert or
+transform that data to be compatible with datasource B, and a third job to
+upload the transformed data to datasource B.
+
+To learn more about workflow design and implementation, see
+[Build & Manage Workflows](/documentation/build/workflows) :::
```
Comment (aleksa-krolls) [inline docs/job-writing-guide.md:32]: @josephjclark all links to other pages need to reference the specific `.md` file in the repo in order in order for the links to not break if we update how we manage doc versions in docusaurus. 

[See this Slack post](https://openfn.slack.com/archives/C06HA4WF35K/p1707811642208039) on the convention we're following for linking to other pages.

File: docs/job-writing-guide.md [hunk 3]
```diff
@@ -0,0 +1,750 @@
...
+  return item.toUpperCase();
+};
+const result = array.map(upperCase);
+console.log(array); // ['a', 'b', 'c'];
+console.log(result); // ['A', 'B', 'C'];
+```
+
+</details>
+
+The `fn()` function, for example, ONLY allows you define a callback. This is
```
Comment (aleksa-krolls) [inline docs/job-writing-guide.md:128]: @josephjclark should we introduce also `common` here? Nowhere in docs so far do we mention that this is a thing... maybe we could also link to [its Adaptor page](https://docs.openfn.org/adaptors/packages/common-docs)?

File: docs/job-writing-guide.md [hunk 4]
```diff
@@ -0,0 +1,750 @@
...
+
+```js
+get('/patients');
+each('$.data.patients', (item, index) => {
+  item.id = `item-${index}`;
+});
+post('/patients', dataValue('patients'));
+```
+
+OpenFn will call your operations in series during workflow execution, ensuring
```
Comment (aleksa-krolls) [inline docs/job-writing-guide.md:204]: @josephjclark should we also link to [this Chaining operations page](https://docs.openfn.org/documentation/build/steps/multiple-operations) and/or edit content on that page?

File: docs/job-writing-guide.md [hunk 5]
```diff
@@ -0,0 +1,750 @@
...
+
+This passes a function into the `post` operator. When the `post()` call actually
+executes, the first thing it'll do is resolve any function into arguments into
+values. It does this by calling the function passing in the latest state, and
+using the return as the value.
+
+These lazy functions are incredibly powerful. Using them effectively is the key
+to writing good OpenFn jobs.
+
+## Iteration and each
```
Comment (aleksa-krolls) [inline docs/job-writing-guide.md:342]: Given the content here - @josephjclark should we revisit [this page on the each() operation](https://docs.openfn.org/documentation/build/steps/each)?

File: docs/job-writing-guide.md [hunk 6]
```diff
@@ -0,0 +1,750 @@
...
+<!--
+I would like to include this BUT fields is not an operation and so works a bit differently
+
+## Object assembly with fields()
+
+Another common pattern is to have to transform one object to another.
+
+The `fields()` operator is a conveient way to do this. `fields()` will assemble an object based on the keys and values you provide. It composes well witth other operations.
+-->
+
```
Comment (aleksa-krolls) [inline docs/job-writing-guide.md:508]: @josephjclark at the end of this section on state, maybe we can link to [this page on initial & final state](https://docs.openfn.org/documentation/build/steps/each)?

File: sidebars-main.js [hunk 7]
```diff
@@ -14,6 +14,7 @@ module.exports = {
         'get-started/implementation-checklist',
       ],
     },
+    'job-writing-guide',
```
Comment (aleksa-krolls) [inline sidebars-main.js:17]: @josephjclark I'm on board with this being more elevated in the sidebar, so long as it links to the more [detailed job-specific docs](https://docs.openfn.org/documentation/build/steps/jobs). Can you please consider how to link the 2 sections?

**Review by aleksa-krolls — CHANGES_REQUESTED** [review]
@josephjclark overall, I think this is a great resource to add and will be super helpful. I left comments to fix links and to consider how we can link to existing docs, so that if a user starts here... they can more easily navigate to other docs pages/dive deeper into specific topics.

**Comment by aleksa-krolls** [comment]
@josephjclark I left initial feedback and then @mtuchi will look at this later today as well

File: docs/job-writing-guide.md (hunk 2 above)
Comment (mtuchi) [reply docs/job-writing-guide.md]: I have pushed a fix for this, See [8aae57e](https://github.com/OpenFn/docs/pull/444/commits/8aae57e5022f4e9c86e1a68fd90f4e6bb8c372de) cc @josephjclark @aleksa-krolls

File: docs/job-writing-guide.md (hunk 4 above)
Comment (mtuchi) [reply docs/job-writing-guide.md:212]: @aleksa-krolls My only problem with the [Chaining operations page](https://docs.openfn.org/documentation/build/steps/multiple-operations) is that it uses `field` and `fields` in that example. And i think that is a pattern we want to deprecate. 

**Two operations approach** 
```
// Prepare a salesforce data model
fn((state) => {
  const { data } = state;

  state.sfMapping = {
    ExternalId: data.form.patient_id,
    Name: data.form.surname,
    "Other Names": data.form.firstName,
    Age__c: data.form.ageInYears,
    Is_Enrolled__c: true,
    Enrollment_Status__c: 3,
  };

  return state;
});

// Upsert patients 
upsert("Patient__c", "ExternalId", (state) => state.sfMapping);
```
**Single operation approach**
```
// Upsert patients
upsert("Patient__c", "ExternalId", ({ data }) => ({
  ExternalId: data.form.patient_id,
  Name: data.form.surname,
  "Other Names": data.form.firstName,
  Age__c: data.form.ageInYears,
  Is_Enrolled__c: true,
  Enrollment_Status__c: 3,
}));
```

I think those examples looks more modern and will cut down explanation on what is `fields`, `filed` and `dataValue`

**Review by mtuchi — COMMENTED** [review]
@josephjclark i have push some improvements in this branch `job-writing-guide`. Please pull if your making changes. Otherwise this is a really good guideline to begin with and we can add improvements as we go. 

Also @aleksa-krolls left some comments, let me know if you need my help addressing them

File: docs/job-writing-guide.md (hunk 3 above)
Comment (josephjclark) [reply docs/job-writing-guide.md:132]: The thing with common is that it should probably be mostly invisible. I think we just need to work out how to list the common functions inside the other adaptors  (which I've raised as #440)

File: docs/job-writing-guide.md (hunk 4 above)
Comment (josephjclark) [reply docs/job-writing-guide.md:212]: Yeah I'm torn on `fields`. Its kinda neat but also quite an awkward pattern. The single operation approach is way neater - and more familiar to Javascript users. And uses re-usable Javascript skills.

I also find that page to be a bit technical and jargon heavy - I'm really trying to strike a gentler tone here.

But we don't want both pages sitting side by side. Let me take a look at reconciling them.

**Comment by josephjclark** [comment]
I've just pushed some fairly drastic changes - someone is gonna be furious with me :sweat: 

image

* I've created a top level section for job writing. This is all CLI and App agnostic
* I've removed a bunch of the old job writing stuff (including some redundancies), and pulled a few extra bits into my docs
* I've created Examples and Snippets sections from old stuff. It's out of scope for this PR to do much with these - but some of those snippets in particular  feel like very bad practice in modern openfn :grimacing: I wonder if we should de-list it for now until we've had a chance to go through and modernise.

File: sidebars-main.js (hunk 7 above)
Comment (josephjclark) [reply sidebars-main.js]: This is really nice I think - makes the docs more discoverable and cohesive 
image

**Comment by mtuchi** [comment]
@josephjclark i think there is a problem when using `a?.b?.c`. The CLI abort with the following error 
```
[CLI] ✘ Failed to compile job job-id-here
[CLI] ✘ Unexpected token (30:50)
[CLI] ♦ Check the syntax of the job expression:
// Some js expressions here ....
[CLI] ✘ Critical error: aborting command
```

**Comment by josephjclark** [comment]
@mtuchi oh my god, thank you for testing that!! Let me take a look.

**Comment by mtuchi** [comment]
> @mtuchi oh my god, thank you for testing that!! Let me take a look.

@josephjclark also the `a.b ?? c` Does not work

**Comment by josephjclark** [comment]
@mtuchi It's all fixed, I'm just prepping the release :)

It's all incidental to this branch anyway. We should document this stuff because I expect job writers to use it.

If it really really refuses to work we'll just comment out the docs for a bit. But it's honestly all done

**Comment by josephjclark** [comment]
Build is (finally) passing. As far as I'm concerned this is ready to merge (squash commit please)

If any typos or tweaks are needed next week, please ago ahead and add them :)

**Review by aleksa-krolls — APPROVED** [review]

**Pull request merged**